File: sdk/samples/02_qos/python/liveliness_manual.py
#!/usr/bin/env python3
# SPDX-License-Identifier: Apache-2.0 OR MIT
# Copyright (c) 2025-2026 naskel.com

"""
HDDS Sample: Liveliness Manual (Python)

Demonstrates MANUAL_BY_PARTICIPANT liveliness - application must
explicitly assert liveliness. Useful for detecting app-level failures.

Usage:
    python liveliness_manual.py        # Subscriber (monitors liveliness)
    python liveliness_manual.py pub    # Publisher (with manual assertion)
"""

import sys
import time

sys.path.insert(0, str(__file__ + "/../../01_basics/python"))

import hdds
from generated.HelloWorld import HelloWorld

LEASE_DURATION_MS = 2000  # 2 second lease
NUM_MESSAGES = 6


def run_publisher(participant):
    """Publish with MANUAL_BY_PARTICIPANT liveliness."""
    qos = hdds.QoS.reliable().liveliness_manual_participant(LEASE_DURATION_MS / 1000.0)
    writer = participant.create_writer("ManualLivenessTopic", qos=qos)

    print(f"Publishing with MANUAL_BY_PARTICIPANT liveliness (lease: {LEASE_DURATION_MS}ms)")
    print("Application must explicitly assert liveliness.\n")

    start = time.monotonic()

    for i in range(NUM_MESSAGES):
        msg = HelloWorld(id=i + 1, message=f"Manual update #{i + 1}")

        # Writing data implicitly asserts liveliness
        writer.write(msg.serialize())

        elapsed = int((time.monotonic() - start) * 1000)
        print(f"  [{elapsed}ms] Published id={msg.id} (liveliness asserted via write)")

        # First 3 messages: normal rate
        # Last 3 messages: slow rate (will miss liveliness)
        if i < 3:
            time.sleep(0.5)  # 500ms - OK
        else:
            print("  (simulating slow processing...)")
            time.sleep(2.5)  # 2.5s - exceeds lease!

    print("\nPublisher done. Some liveliness violations occurred.")


def run_subscriber(participant):
    """Monitor MANUAL_BY_PARTICIPANT liveliness."""
    qos = hdds.QoS.reliable().liveliness_manual_participant(LEASE_DURATION_MS / 1000.0)
    reader = participant.create_reader("ManualLivenessTopic", qos=qos)

    waitset = hdds.WaitSet()
    waitset.attach(reader)

    print(f"Monitoring MANUAL_BY_PARTICIPANT liveliness (lease: {LEASE_DURATION_MS}ms)...")
    print("Writer must assert liveliness explicitly (by writing).\n")

    received = 0
    liveliness_changed = 0
    start = time.monotonic()
    last_msg = start

    while received < NUM_MESSAGES or liveliness_changed < 3:
        if waitset.wait(timeout_secs=LEASE_DURATION_MS / 1000.0):
            while True:
                data = reader.take()
                if data is None:
                    break

                msg = HelloWorld.deserialize(data)
                now = time.monotonic()
                elapsed = int((now - start) * 1000)
                delta = int((now - last_msg) * 1000)

                status = " [LIVELINESS WAS LOST]" if (delta > LEASE_DURATION_MS and received > 0) else ""

                print(f"  [{elapsed}ms] Received id={msg.id} (delta={delta}ms){status}")

                last_msg = now
                received += 1
        else:
            now = time.monotonic()
            elapsed = int((now - start) * 1000)
            since_last = int((now - last_msg) * 1000)

            if since_last > LEASE_DURATION_MS and received > 0:
                print(f"  [{elapsed}ms] LIVELINESS LOST! (no assertion for {since_last}ms)")
                liveliness_changed += 1

            if liveliness_changed >= 3:
                break

    print()
    print("-" * 60)
    print(f"Summary: {received} messages, {liveliness_changed} liveliness events detected")
    print("MANUAL liveliness requires explicit app-level assertion.")
    print("-" * 60)


def main():
    is_publisher = len(sys.argv) > 1 and sys.argv[1] == "pub"

    hdds.logging.init(hdds.LogLevel.INFO)

    print("=" * 60)
    print("Liveliness Manual Demo")
    print("QoS: MANUAL_BY_PARTICIPANT - app must assert liveliness")
    print("=" * 60)

    with hdds.Participant("LivelinessManualDemo") as participant:
        if is_publisher:
            run_publisher(participant)
        else:
            run_subscriber(participant)


if __name__ == "__main__":
    main()

File: sdk/samples/02_qos/cpp/liveliness_manual.cpp
// SPDX-License-Identifier: Apache-2.0 OR MIT
// Copyright (c) 2025-2026 naskel.com

/**
 * HDDS Sample: Liveliness Manual (C++)
 *
 * Demonstrates MANUAL_BY_PARTICIPANT liveliness - application must
 * explicitly assert liveliness. Useful for detecting app-level failures.
 *
 * Usage:
 *     ./liveliness_manual        # Subscriber (monitors liveliness)
 *     ./liveliness_manual pub    # Publisher (with manual assertion)
 */

#include <hdds.hpp>
#include <iostream>
#include <thread>
#include <chrono>
#include <cstring>

#include "generated/HelloWorld.hpp"

using namespace hdds_samples;
using namespace std::chrono_literals;
using Clock = std::chrono::steady_clock;

constexpr auto LEASE_DURATION = 2000ms;  /* 2 second lease */
constexpr int NUM_MESSAGES = 6;

void run_publisher(hdds::Participant& participant) {
    /* Create writer with MANUAL_BY_PARTICIPANT liveliness */
    auto qos = hdds::QoS::reliable()
        .liveliness_manual_participant(LEASE_DURATION);

    auto writer = participant.create_writer<HelloWorld>("ManualLivenessTopic", qos);

    std::cout << "Publishing with MANUAL_BY_PARTICIPANT liveliness (lease: "
              << LEASE_DURATION.count() << "ms)\n";
    std::cout << "Application must explicitly assert liveliness.\n\n";

    auto start = Clock::now();

    for (int i = 0; i < NUM_MESSAGES; i++) {
        HelloWorld msg(i + 1, "Manual update #" + std::to_string(i + 1));
        /* Writing data implicitly asserts liveliness */
        writer.write(msg);

        auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(
            Clock::now() - start).count();
        std::cout << "  [" << elapsed << "ms] Published id=" << msg.id
                  << " (liveliness asserted via write)\n";

        /* First 3 messages: normal rate
         * Last 3 messages: slow rate (will miss liveliness) */
        if (i < 3) {
            std::this_thread::sleep_for(500ms);  /* OK */
        } else {
            std::cout << "  (simulating slow processing...)\n";
            std::this_thread::sleep_for(2500ms);  /* Exceeds lease! */
        }
    }

    std::cout << "\nPublisher done. Some liveliness violations occurred.\n";
}

void run_subscriber(hdds::Participant& participant) {
    /* Create reader with MANUAL_BY_PARTICIPANT liveliness */
    auto qos = hdds::QoS::reliable()
        .liveliness_manual_participant(LEASE_DURATION);

    auto reader = participant.create_reader<HelloWorld>("ManualLivenessTopic", qos);

    hdds::WaitSet waitset;
    waitset.attach(reader.get_status_condition());

    std::cout << "Monitoring MANUAL_BY_PARTICIPANT liveliness (lease: "
              << LEASE_DURATION.count() << "ms)...\n";
    std::cout << "Writer must assert liveliness explicitly (by writing).\n\n";

    int received = 0;
    int liveliness_changed = 0;
    auto start = Clock::now();
    auto last_msg = start;

    while (received < NUM_MESSAGES || liveliness_changed < 3) {
        if (waitset.wait(LEASE_DURATION)) {
            while (auto msg = reader.take()) {
                auto now = Clock::now();
                auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(
                    now - start).count();
                auto delta = std::chrono::duration_cast<std::chrono::milliseconds>(
                    now - last_msg).count();

                std::string status = (delta > LEASE_DURATION.count() && received > 0)
                    ? " [LIVELINESS WAS LOST]" : "";

                std::cout << "  [" << elapsed << "ms] Received id=" << msg->id
                          << " (delta=" << delta << "ms)" << status << "\n";

                last_msg = now;
                received++;
            }
        } else {
            auto now = Clock::now();
            auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(
                now - start).count();
            auto since_last = std::chrono::duration_cast<std::chrono::milliseconds>(
                now - last_msg).count();

            if (since_last > LEASE_DURATION.count() && received > 0) {
                std::cout << "  [" << elapsed << "ms] LIVELINESS LOST! (no assertion for "
                          << since_last << "ms)\n";
                liveliness_changed++;
            }

            if (liveliness_changed >= 3) {
                break;
            }
        }
    }

    std::cout << "\n" << std::string(60, '-') << "\n";
    std::cout << "Summary: " << received << " messages, "
              << liveliness_changed << " liveliness events detected\n";
    std::cout << "MANUAL liveliness requires explicit app-level assertion.\n";
    std::cout << std::string(60, '-') << "\n";
}

int main(int argc, char** argv) {
    bool is_publisher = (argc > 1 && std::strcmp(argv[1], "pub") == 0);

    try {
        hdds::logging::init(hdds::LogLevel::Warn);

        std::cout << std::string(60, '=') << "\n";
        std::cout << "Liveliness Manual Demo\n";
        std::cout << "QoS: MANUAL_BY_PARTICIPANT - app must assert liveliness\n";
        std::cout << std::string(60, '=') << "\n";

        hdds::Participant participant("LivelinessManualDemo");

        if (is_publisher) {
            run_publisher(participant);
        } else {
            run_subscriber(participant);
        }

    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << "\n";
        return 1;
    }

    return 0;
}
